File: 0051-n-queens/0051-n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        ans = []
        board = [["."]*n for i in range(n)]
        self.solve(0,n,board,ans)
        return ans
    def solve(self,col,n,board,ans):
        if col == n:
            x = ["".join(i) for i in board]
            ans.append(x)
            print(board)
            return
        for i in range(0,n):
            if self.safe(i,col,n,board):
                board[i][col] = "Q"
                self.solve(col+1,n,board,ans)
                board[i][col] = "."
        return
    def safe(self,row,col,n,board):
        r = row
        c = col
        while r >=0 and c >=0:
            if board[r][c] == "Q":
                return False
            r-=1
            c-=1
        r = row
        c = col
        while c>=0:
            if board[r][c] =="Q":
                return False
            c-=1
        r = row
        c = col

        while r<n and c>=0:
            if board[r][c] == "Q":
                return False
            c-=1
            r+=1
        return True

        

File: 0051-n-queens/0051-n-queens.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
       boolean [][] board = new boolean[n][n];
       
       
       return queens(board,0);

    }
    static List<List<String>>  queens (boolean[][] board,int row)
    {
        if (row==board.length)
        {
            return display(board);

        }
        List<List<String>> ans = new ArrayList<>();
        for (int i = 0 ; i < board[0].length;i++)
        {
            
            if (isSafe(board,row,i))
            {
            board[row][i] = true;
            ans.addAll(queens(board,row+1));
            board[row][i] = false;
            }
            
        }
        return ans;
    }
    static boolean isSafe(boolean[][] board,int row,int col)
    {
        for (int i=0;i<=row;i++)
        {
            if (board[i][col])
            {
                return false;
            }
        }
        int minleft = Math.min(row,col);
        for (int i = 1;i<=minleft;i++)
        {
            if (board[row-i][col-i])
            {
                return false;
            }
        }
        int minright = Math.min(row,board.length-col-1);
        for (int i = 1;i<=minright;i++)
        {
            if (board[row-i][col+i])
            {
                return false;
            }
        }
        return true;
    }
    static List<List<String>> display (boolean[][] board)
    {   
        List<List<String>> ans = new ArrayList<>();
        List<String> sol = new ArrayList<>();
        for (int i= 0; i < board.length; i++)
        {   String s = "";
            for (int j = 0; j<board[0].length;j++)
            {
                if (board[i][j])
                {
                    s+="Q";
                    
                }
                else
                {
                    s+=".";
                }
            }
            sol.add(s);
        }
        ans.add(sol);
        return ans;
    }
}